source/common/stats/deferred_creation.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+
+#include "source/common/common/cleanup.h"
+#include "source/common/common/thread.h"
+#include "source/common/stats/symbol_table.h"
+#include "source/common/stats/utility.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ * This is used when
+ * :ref:`enable_deferred_creation_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_deferred_creation_stats>` is enabled.
+ */
+template <typename StatsStructType>
+class DeferredCreation : public DeferredCreationCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  DeferredCreation(const typename StatsStructType::StatNameType& stat_names,
+                   Stats::ScopeSharedPtr scope)
+      : initialized_([&scope]() -> Gauge& {
+          Stats::StatNamePool pool(scope->symbolTable());
+          return Stats::Utility::gaugeFromElements(
+              *scope, {pool.add(StatsStructType::typeName()), pool.add("initialized")},
+              Stats::Gauge::ImportMode::HiddenAccumulate);
+        }()),
+        ctor_([stats_scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          initialized_.inc();
+          // Reset ctor_ to save some RAM.
+          Cleanup reset_ctor([&] { ctor_ = nullptr; });
```

Comment:
Usually we don't capture `[&]` in prod code. Is it sufficient to capture `this`? If not can you add comments?

Reply:
we want a copy of the the scope shard ptr. 
updated.